secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/consent/ConsentRequestAccessAuthorisationFilter.java

```diff
@@ -0,0 +1,181 @@
+import org.forgerock.util.promise.Promises;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.forgerock.sapi.gateway.fapi.FAPIUtils;
+
+/**
+ * Filter that protects access to a consent request by validating that an end user browser session belongs to the
+ * same user that is specified in the signed consent request JWT (sent from AM). This can be used in filter chains
+ * which protect Remote Consent flows.
+ * <p>
+ * This filter depends on the {@link SsoTokenContext} and {@link JwtValidationContext}, therefore it must be installed
+ * after filters which add these contexts.
+ * <p>
+ * The {@link SsoTokenContext} is used to determine the user that is logged in, by inspecting the session "uid".
+ * <p>
+ * The {@link JwtValidationContext} is used to determine the user that owns the consent request, by inspecting the
+ * "username" claim. It is assumed that consent request JWT has been fully validated, that the signature has been verified and that
+ * the "exp", "iat", "iss" and "aud" claims have all been validated.
+ * <p>
+ * If the SSO user matches the consent user then this filter passes the request on to the next handler in the chain.
+ * Otherwise, this filter responds with HTTP 401.
+ * <p>
+ * If any exceptions are raised when extracting the user data from the contexts then this filter responds with HTTP 500.
+ * This is because an exception indicates either an error in the IG configuration or AM configuration; there is no action
+ * that an end user can take to resolve the issue.
+ */
+public class ConsentRequestAccessAuthorisationFilter implements Filter {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ConsentRequestAccessAuthorisationFilter.class);
+    static final String DEFAULT_CONSENT_REQUEST_USER_ID_CLAIM = "username";
+    static final String DEFAULT_SSO_TOKEN_USER_ID_KEY = "uid";
+
+    private final String consentRequestUserIdClaim;
+    private final String ssoTokenUserIdKey;
+
+    /**
+     * Handles exceptions raised by the business logic of this filter and converts them into HTTP Response objects
+     */
+    @FunctionalInterface
+    interface ExceptionHandler {
+        Response onException(Context context, Exception ex);
+    }
+
+    /**
+     * @return ExceptionHandler instance which logs the exception and returns a HTTP 500 response. All exceptions in
+     * this filter are unexpected, there is no action that the end user can take to resolve issues. If an exception
+     * occurs then it indicates a misconfiguration in a ForgeRock application.
+     */
+    static ExceptionHandler createDefaultExceptionHandler() {
+        return (ctxt, ex) -> {
+            LOGGER.warn("(" + FAPIUtils.getFapiInteractionIdForDisplay(ctxt) + ") Failed to get userId data required to do authorisation check", ex);
+            return new Response(Status.INTERNAL_SERVER_ERROR);
+        };
+    }
+
+    private final ExceptionHandler exceptionHandler;
```

Comment:
This is a hook for testing purposes, it allows the unit tests to supply an ExceptionHandler instance which captures the exceptions passed to it. We can then do validation on these exceptions to ensure that a HTTP 500 response was triggered by a particular exception.

The 500 response should only occur if AM or IG change their impl of either the consent_request JWT or SsoToken format. There is intentionally no information returned to the end user as there is no action they can reasonably take, we want to avoid leaking any information at this point as another reason for a failure is someone attempting to hack this flow.